### PR TITLE
[trainer] fix: fail closed on incomplete agent-loop rollouts

### LINKED ISSUE #6437: [trainer] main_ppo_sync can silently train on partial rollout batches after agent-loop failures
### System Info

Not environment-specific. This is a source-level behavior in `verl/trainer/main_ppo_sync.py` on current main.

### Information

- [x] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

This is a source-level issue in `verl.trainer.main_ppo_sync` when using the TransferQueue/agent-loop path.

Relevant code path:

1. `AgentLoopManagerTQ.generate_sequences()` registers one root key per prompt in the local `ReplayBuffer` with `status="running"`.

2. `AgentLoopWorkerTQ._run_prompt()` launches `rollout.n` session tasks for that prompt. When those tasks finish, it writes the root prompt key as `status="finished"`. If an exception escapes, it writes the root prompt key as `status="failure"`.

3. The actual trajectory rows are not written under the root prompt key. `_agent_loop_postprocess()` writes child keys of the form `{uid}_{session_id}_{index}` with `status="success"`.

4. If a session returns an empty output list, `_agent_loop_postprocess()` logs a warning and returns without writing a success child row.

5. `ReplayBuffer.sample(partition_id=..., global_steps=...)` waits only while it sees a key with `status="running"`. It appends keys with `status="success"`, but ignores other terminal statuses such as `finished` and `failure`.

6. Therefore, once no root marker remains `running`, `ReplayBuffer.sample()` can return a `KVBatchMeta` containing only the successful child rows, even if some requested sessions failed or produced no output.

A concrete trigger shape:

- Use `main_ppo_sync`.
- Use agent-loop rollout through TransferQueue.
- Set `actor_rollout_ref.rollout.n > 1`.
- Have one prompt/session fail or return an empty output list.
- The trainer can continue the PPO step with fewer rollout rows than requested.

This can silently reduce the effective rollout count for the step.

### Expected behavior

`main_ppo_sync` should make the partial-rollout policy explicit.

At minimum, when sampling by `global_steps`, the trainer should not silently treat missing child rollout rows as a complete full batch.

Possible acceptable behaviors:

1. Fail closed by default if any prompt/session reaches `failure` or if a finished prompt produced fewer child success rows than expected.

2. Or intentionally allow partial rollout batches, but expose that policy clearly through configuration and metrics, such as:
   - requested rollout rows
   - failed prompt/session count
   - validation denominator changes

For GRPO-style training, missing sibling rollouts can change the per-prompt group statistics, so silently proceeding with only surviving rows is difficult to distinguish from a healthy full batch.

### PR DESCRIPTION
**Pull request by TimurTaepov**
### What does this PR do?

Fixes #6437.

This PR makes `main_ppo_sync` fail closed when agent-loop rollout batches finish with missing rollout sessions.

Previously, `ReplayBuffer.sample()` waited only for `status="running"` markers to disappear, collected `status="success"` rows, and silently ignored terminal root statuses such as `finished` and `failure`. This could let training continue on a partial rollout group after an agent-loop failure or empty output.

This PR:
- records the expected rollout session count on each root prompt marker;
- detects terminal prompt failures;
- detects `finished` prompts with fewer successful rollout sessions than expected;
- cleans up sampled root prompt markers together with child output rows;
- raises a clear error instead of returning a partial `KVBatchMeta`;
- adds CPU regression tests.

### Checklist Before Starting

- [x] Search for similar PRs. Paste at least one query link here:
  - https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+6437+in%3Abody
  - https://github.com/verl-project/verl/pulls?q=is%3Apr+is%3Aopen+ReplayBuffer+partial+rollout+failure+main_ppo_sync
- [x] Format the PR title as `[{modules}] {type}: {description}`

### Test

```bash
.venv/bin/python -m pytest tests/trainer/test_main_ppo_sync_replay_buffer_on_cpu.py -q
# 9 passed

.venv/bin/python -m pytest tests/trainer/test_multi_trajectories_advantage.py -q
# 2 passed

.venv/bin/pre-commit run ruff --files verl/trainer/main_ppo_sync.py tests/trainer/test_main_ppo_sync_replay_buffer_on_cpu.py
# Passed

.venv/bin/pre-commit run ruff-format --files verl/trainer/main_ppo_sync.py tests/trainer/test_main_ppo_sync_replay_buffer_on_cpu.py
# Passed

git diff --check
# Passed
```

During commit, the local `check-naming-conventions` pre-commit hook was skipped because it scans `.venv` and false-positives on Ray's installed `ServeRLlibRLModule`. Running the same grep check with `.venv` excluded found no matches in the repository files.

### API and Usage Example

No public API or usage changes.

### Design & Code Changes

`AgentLoopManagerTQ.generate_sequences()` now records `expected_rollouts` on each root prompt marker before dispatching agent-loop workers.

`ReplayBuffer.sample()` now:
- waits while any matching prompt is still `running`;
- collects successful child output rows as before;
- counts unique successful rollout sessions per prompt using child keys formatted as `{uid}_{session_id}_{index}`;
- carries the sampled root prompt keys in `KVBatchMeta.extra_info` so cleanup removes both root markers and child rows;
- raises `RuntimeError` if a root prompt reaches `failure`;
- raises `RuntimeError` if a `finished` root prompt has fewer successful sessions than `expected_rollouts`.

The child-key parser uses `rsplit("_", 2)` because `uid` can contain underscores. The check counts unique sessions rather than output rows, because one agent-loop session can emit multiple output rows.

### Checklist Before Submitting

- [x] Read the [Contribute Guide](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update [the documentation](https://github.com/verl-project/verl/tree/main/docs).
- [x] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl/tree/main/.github/workflows) to cover all the code.
- [ ] Once your PR is ready for CI, send a message in the `ci-request` channel.
- [x] This PR is not related to the `recipe` submodule.

AI assistance disclosure: this PR was developed with AI assistance. I reviewed the changed lines and ran the tests above.

### CODE REVIEW

File: verl/trainer/main_ppo_sync.py [hunk 1]
```diff
@@ -279,18 +313,47 @@ def sample(self, partition_id: str, global_steps: int = None, batch_size: int =
...
                         else:
-                            logger.debug(f"Unknown status {tag['status']} for key {key}")
+                            expected = tag.get("expected_rollouts")
+                            if expected is not None:
+                                if status == "finished":
+                                    expected_rollouts[key] = int(expected)
+                                else:
+                                    failures[key] = tag
+                            else:
+                                logger.debug(f"Unknown status {status} for key {key}")
```
[the remark this answers is not in the document]
Comment (TimurTaepov) [reply verl/trainer/main_ppo_sync.py]: Reviewed this point. I separated `finished` and `failure` handling so failures are always captured, and added a regression test for a failed prompt without `expected_rollouts`

**Comment by TimurTaepov** [comment]
The `cpu_unit_tests.yml` workflow parse error appears unrelated to this PR. This branch does not modify workflow files; the same invalid step is present on latest `upstream/main` at `.github/workflows/cpu_unit_tests.yml:98`, where `Check final pip list` is missing a `run:` key.

I’ll leave this PR scoped to the trainer fix unless maintainers prefer otherwise.

**Comment by Baiyu-Su** [comment]
Thanks for the PR. The fail-closed direction matches the issue, and counting unique `{uid}_{session_id}` sessions rather than output rows looks right for multi-output agent loops.

I think there is still one cleanup edge case from the original issue: root prompt markers are not removed after a successful sample. Both training and validation cleanup currently clear/remove only `batch.keys`, which are child success rows. The root `uid` markers remain in the replay buffer / TransferQueue metadata.

This becomes observable with the new completeness check in validation. `_validate()` iterates over all validation dataloader batches using the same `self.global_steps`. After the first val batch succeeds, cleanup removes only the child success rows. The root marker from that batch remains as `finished` with `expected_rollouts`. On the next val batch at the same `global_steps`, `ReplayBuffer.sample(partition_id="val", global_steps=...)` can see the stale finished root, count zero successful sessions for that old `uid`, and raise a false incomplete-rollout error.

Could we include root-marker cleanup in this PR, or at least add a regression test for two complete validation-style batches at the same `global_steps` with cleanup between them? I think the sampled batch needs a way to carry the root keys used for completeness accounting so cleanup can remove both root markers and child output rows.

**Comment by TimurTaepov** [comment]
Hi @Baiyu-Su, thanks, you're right. The sampled batch only carried child success rows, so cleanup could leave finished root markers behind.

I updated `ReplayBuffer.sample()` to carry the sampled root prompt markers in `KVBatchMeta.extra_info["rollout_root_keys"]`, and updated both validation and training cleanup to clear/remove child rows plus those root markers.

I also added a regression test for two complete validation-style batches at the same `global_steps` with cleanup between them.

**Comment by TimurTaepov** [comment]
Gentle ping on this PR. I addressed the root-marker cleanup edge case raised by @Baiyu-Su , updated the train/validation cleanup path, and added the validation-style regression test.

The only failing check I’m aware of is the unrelated `cpu_unit_tests.yml` workflow parse error, which appears to be present on current `upstream/main` as well. Happy to adjust anything else needed for review.
